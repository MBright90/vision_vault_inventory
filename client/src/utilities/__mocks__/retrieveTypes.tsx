import React from "react";

export default jest.fn(async (setFilterOptions, setFilterIdArr, selected) => {
    const data = [
        { _id: 'type1', name: 'Type 1' },
        { _id: 'type2', name: 'Type 2' },
    ];

    const filterNodes = data.map(type => (
        <option key={type._id} data-id={type._id} selected={selected === type.name}>
            {type.name}
        </option>
    ));
    setFilterOptions(filterNodes);

    if (setFilterIdArr) {
        const filterIds = data.map(type => ({ name: type.name, _id: type._id }));
        setFilterIdArr([{ name: 'all', _id: 'all' }, ...filterIds]);
    }
});