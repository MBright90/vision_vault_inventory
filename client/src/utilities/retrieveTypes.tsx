import React from "react";
import capitalize from "./capitalize";

import { type productType_type } from "@custom_types/types";
import endpoint from "./endpoint";

interface FilterIdSet{
    name: string,
    _id: string
}

const retrieveTypes = async (
    setFilterOptions: (nodes: React.ReactNode[]) => void,
    setFilterIdArr: ((idSet: FilterIdSet[]) => void) | null,
    selected: string | null = null,
    ): Promise<void> => {
        let formattedSelected: string = "";
        if (selected !== null) {
            formattedSelected = capitalize(selected);
        }
    let data: productType_type[] = [];
    try {
        const response = await fetch(`${endpoint}/types/`);
        data = await response.json();
        // Create ReactNode arr
        const filterNodes: React.ReactNode[] = data.map((type: productType_type) => {
            if (capitalize(formattedSelected) === type.name) {
                return (<option key={type._id} data-id={type._id} selected={true}>
                    {capitalize(type.name)}
                </option>);
            } else {
                return (<option key={type._id} data-id={type._id}>
                    {capitalize(type.name)}
                </option>);
            }
        });
        setFilterOptions(filterNodes);

        // Create filterId arr
        if (setFilterIdArr !== null) {
            const filterIds: FilterIdSet[] = data.map((type: productType_type) => {
                return { name: type.name, _id: type._id };
            });
            setFilterIdArr([{ name: 'all', _id: 'all' }, ...filterIds]);
        }
    } catch (err) {
        console.log(err);
    }
};

export default retrieveTypes;
