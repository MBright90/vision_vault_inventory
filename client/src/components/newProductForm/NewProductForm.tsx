import React, { useEffect, useState } from "react";

import retrieveTypes from "@utilities/retrieveTypes";
import { type FilterIdSet } from "@custom_types/types";

import style from './NewProductForm.module.scss';

const NewProductForm: React.FC = () => {
    const [typeIdArr, setTypeIdArr] = useState<FilterIdSet[]>([]);
    const [typeOptions, setTypeOptions] = useState<React.ReactNode[]>([]);

    const [currentName, setCurrentName] = useState<string>('');
    const [currentDescription, setCurrentDescription] = useState<string>('');
    const [currentStock, setCurrentStock] = useState<number>(0);
    const [currentType, setCurrentType] = useState<string>('Select type...');

    useEffect(() => {
        void(retrieveTypes(setTypeOptions, setTypeIdArr));
    }, []);

    useEffect(() => { // to remove
        console.log(currentName);
        console.log(currentDescription);
        console.log(currentStock);
        console.log(currentType);
        console.log(typeIdArr);
    }, [currentType]);

    function handleNameChange(e: React.ChangeEvent<HTMLInputElement>): void {
        setCurrentName(e.target.value);
    }

    function handleDescriptionChange(e: React.ChangeEvent<HTMLTextAreaElement>): void {
        setCurrentDescription(e.target.value);
    }

    function handleStockChange(e: React.ChangeEvent<HTMLInputElement>): void {
        const { value } = e.target;
        if (typeof value === 'number' && value >= 0) setCurrentStock(value);
    }

    function handleTypeChange(e: React.ChangeEvent<HTMLSelectElement>): void {
        setCurrentType(e.target.value.toLowerCase());
    }

    return (<div className={style.formContainer}>
        <h3>Add Product</h3>
        <form action="" className={style.newProductForm}>

            <div className={style.inputContainer}>
                <label htmlFor="name-input">Product Name</label>
                <input id="name-input" name="name-input" type="text" onChange={(e) => { handleNameChange(e); }} required />
            </div>

            <div className={style.inputContainer}>
                <label htmlFor="description-input">Description</label>
                <textarea id="description-input" name="description-input" onChange={(e) => { handleDescriptionChange(e); }} required />
            </div>

            <div className={style.inputContainer}>
                <label htmlFor="stock-input">Initial stock</label>
                <input id="stock-input" name="stock-input" type="number" max={100} min={0} onChange={(e) => { handleStockChange(e); }} required />
                <p className={style.inputInfo}>
                    Max 100
                </p>
            </div>

            <div className={style.inputContainer}>
                <label htmlFor="type-input">Type</label>
                <select id="type-input" name="type-input" onChange={(e) => { handleTypeChange(e); }} required>
                    {typeOptions}
                </select>
                {/* <p className={style.inputInfo}>
                    Separate genres with a comma
                </p> */}
            </div>

            <div className={style.inputContainer}>
                <label htmlFor="genres-input">Genres</label>
                <input id="genres-input" name="genres-input" type="text" required />
                <p className={style.inputInfo}>
                    Separate genres with a comma
                </p>
            </div>

            <button className={style.formSubmitBtn} type="submit">Add Product</button>
        </form>
    </div>
    );
};

export default NewProductForm;
