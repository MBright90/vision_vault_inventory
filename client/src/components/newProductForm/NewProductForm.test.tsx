import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom"
import NewProductForm from "./NewProductForm";

import retrieveTypes from "@utilities/retrieveTypes";

jest.mock('@components/modals/confirmModal/ConfirmModal', () => ({
    __esModule: true,
    default: (props: { children: React.ReactNode[], closeModal: () => {} }) => {
        return (
            <div data-testid="mock-confirm-modal">
                {props.children}
            </div>
        )
    }
}));

jest.mock('@utilities/retrieveTypes', () => jest.fn());

beforeAll(async () => {
    global.fetch = jest.fn();
});

beforeEach(() => {
    jest.resetAllMocks();
});

test('matches snapshot', () => {
    const { container } = render(<NewProductForm />);
    
    expect(container).toMatchSnapshot();
});

test('calls retrieveTypes on component mount', () => {
    render(<NewProductForm />);

    expect(retrieveTypes).toHaveBeenCalledTimes(1);
});

test('renders form with initial empty values', () => {
    render(<NewProductForm />);

    expect(screen.getByLabelText(/Product Name/i)).toHaveValue('');
    expect(screen.getByLabelText(/Description/i)).toHaveValue('');
    expect(screen.getByLabelText(/Price/i)).toHaveValue(0);
    expect(screen.getByLabelText(/Initial Stock/i)).toHaveValue(0);
    expect(screen.getByLabelText(/Type/i)).toHaveValue('Select Type...');
    expect(screen.getByLabelText(/Genres/i)).toHaveValue('');
});

test('updates the product name input value when typed into', () => {
    render(<NewProductForm />);
    
    const nameInput = screen.getByLabelText(/Product Name/i);
    fireEvent.change(nameInput, { target: { value: 'Test Product Name' } });
    expect(nameInput).toHaveValue('Test Product Name');
});

test('updates the product description input value when typed into', () => {
    render(<NewProductForm />);

    const descriptionInput = screen.getByLabelText(/Description/i);
    fireEvent.change(descriptionInput, { target: { value: 'Test Product Description' } });
    expect(descriptionInput).toHaveValue('Test Product Description');
});

test('updates the product price input value when typed into', () => {
    render(<NewProductForm />)

    const priceInput = screen.getByLabelText(/Price/i);
    fireEvent.change(priceInput, { target: { value: '100' } });
    expect(priceInput).toHaveValue(100);
});

test('updates the product initial stock input value when typed into', () => {
    render(<NewProductForm />)

    const stockInput = screen.getByLabelText(/Price/i);
    fireEvent.change(stockInput, { target: { value: '200' } });
    expect(stockInput).toHaveValue(200);
});

test('updates the product genres input value when typed into', () => {
    render(<NewProductForm />)

    const genresInput = screen.getByLabelText(/Genres/i);
    fireEvent.change(genresInput, { target: { value: 'test genre one, test genre two'} });
    expect(genresInput).toHaveValue('test genre one, test genre two');
});

test('form is submitted when submit button is clicked', async () => {
    render(<NewProductForm />);

    const submitButton = screen.getByRole('button');
    fireEvent.click(submitButton);

    await waitFor(() => {
        expect(fetch).toHaveBeenCalled();
    });
});

test('upon successful form submission, a modal shows success message', async () => {
    global.fetch = jest.fn((): Promise<Response> => 
        Promise.resolve({
            ok: true,
            json: async () => {},
        } as Response)
    );
    
    render(<NewProductForm />);

    const submitButton = screen.getByRole('button');
    fireEvent.click(submitButton);
    expect(fetch).toHaveBeenCalled();

    await waitFor(() => {
        expect(screen.getByTestId('mock-confirm-modal')).toBeInTheDocument();
        expect(screen.getByText('Product added successfully')).toBeInTheDocument();
    });
});

test('upon unsuccessful form submission, a modal shows an error message', async () => {
    global.fetch = jest.fn((): Promise<Response> => 
        Promise.resolve({
            ok: false,
            json: async (): Promise<{ err: string }> => Promise.resolve({
                err: 'test error message'
            }),
        } as Response)
    );
    
    render(<NewProductForm />);

    const submitButton = screen.getByRole('button');
    fireEvent.click(submitButton);
    expect(fetch).toHaveBeenCalled();

    await waitFor(() => {
        expect(screen.getByTestId('mock-confirm-modal')).toBeInTheDocument();
        expect(screen.getByText(/Error adding product/i)).toBeInTheDocument();
        expect(screen.getByText(/test error message/i)).toBeInTheDocument();
    });
});

test('handles component unmounting without errors', () => {
    const { unmount } = render(<NewProductForm />);

    expect(() => unmount()).not.toThrow();
});
