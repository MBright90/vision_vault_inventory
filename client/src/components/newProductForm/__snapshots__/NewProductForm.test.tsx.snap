// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <main>
    <h3>
      Add Product
    </h3>
    <form
      action=""
    >
      <div>
        <label
          for="name-input"
        >
          Product Name
        </label>
        <input
          id="name-input"
          name="name-input"
          required=""
          type="text"
        />
        <p>
          Required
        </p>
        <p>
          2 - 150 characters
        </p>
      </div>
      <div>
        <label
          for="description-input"
        >
          Description
        </label>
        <textarea
          id="description-input"
          name="description-input"
          required=""
        />
      </div>
      <div>
        <label
          for="price-input"
        >
          Price
        </label>
        <input
          id="price-input"
          min="0"
          name="price-input"
          required=""
          step="0.01"
          type="number"
          value="0"
        />
        <p>
          Required
        </p>
        <p>
          Must be more than 0.00
        </p>
      </div>
      <div>
        <label
          for="stock-input"
        >
          Initial stock
        </label>
        <input
          id="stock-input"
          max="100"
          min="0"
          name="stock-input"
          required=""
          step="1"
          type="number"
          value="0"
        />
        <p>
          Required
        </p>
      </div>
      <div>
        <label
          for="type-input"
        >
          Type
        </label>
        <select
          id="type-input"
          name="type-input"
          required=""
        >
          <option>
            Select Type...
          </option>
        </select>
        <p>
          Required
        </p>
      </div>
      <div>
        <label
          for="genres-input"
        >
          Genres
        </label>
        <input
          id="genres-input"
          name="genres-input"
          required=""
          type="text"
        />
        <p>
          Minimum of one required
        </p>
        <p>
          Separate genres with a comma
        </p>
      </div>
      <button
        type="submit"
      >
        Add Product
      </button>
    </form>
  </main>
</div>
`;
